File: src/main/java/name/nkid00/rcutil/helper/PosHelper.java
package name.nkid00.rcutil.helper;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;

import name.nkid00.rcutil.util.Blocks;
import name.nkid00.rcutil.util.TargetBlockPos;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.util.math.Vec3f;
import net.minecraft.util.math.Vec3i;

public class PosHelper {
    public static Vec3i toVec3i(BlockPos v) {
        if (v == null) {
            return null;
        }
        return new Vec3i(v.getX(), v.getY(), v.getZ());
    }

    public static BlockPos toBlockPos(Vec3i v) {
        if (v == null) {
            return null;
        }
        return new BlockPos(v);
    }

    public static Vec3i toVec3i(Vec3f v) {
        if (v == null) {
            return null;
        }
        return new Vec3i(v.getX(), v.getY(), v.getZ());
    }

    public static Vec3f toVec3f(Vec3i v) {
        if (v == null) {
            return null;
        }
        return new Vec3f(v.getX(), v.getY(), v.getZ());
    }

    public static Vec3i fromJson(JsonElement v) {
        return fromJson(v.getAsJsonArray());
    }

    public static Vec3i fromJson(JsonArray v) {
        return new Vec3i(v.get(0).getAsInt(), v.get(1).getAsInt(), v.get(2).getAsInt());
    }

    public static Vec3f center(BlockPos v) {
        return new Vec3f(v.getX() + 0.5f, v.getY() + 0.5f, v.getZ() + 0.5f);
    }

    public static BlockPos copy(BlockPos v) {
        if (v == null) {
            return null;
        }
        return v.mutableCopy();
    }

    public static TargetBlockPos copy(TargetBlockPos v) {
        if (v == null) {
            return null;
        }
        return v.copy();
    }

    public static Vec3i copy(Vec3i v) {
        if (v == null) {
            return null;
        }
        return new Vec3i(v.getX(), v.getY(), v.getZ());
    }

    public static Vec3f copy(Vec3f v) {
        if (v == null) {
            return null;
        }
        return new Vec3f(v.getX(), v.getY(), v.getZ());
    }

    public static Blocks copy(Blocks v) {
        if (v == null) {
            return null;
        }
        return new Blocks(v.first(), v.increment(), v.size());
    }

    public static Vec3i scale(Vec3i v, int factor) {
        if (factor == 1) {
            return copy(v);
        }
        return new Vec3i(v.getX() * factor, v.getY() * factor, v.getZ() * factor);
    }

    public static Vec3i divide(Vec3i v, int divisor) {
        if (divisor == 1) {
            return copy(v);
        }
        return new Vec3i(v.getX() / divisor, v.getY() / divisor, v.getZ() / divisor);
    }

    public static Vec3f scale(Vec3f v, float factor) {
        if (DataHelper.isFloatEqual(factor, 1.0f)) {
            return copy(v);
        }
        return new Vec3f(v.getX() * factor, v.getY() * factor, v.getZ() * factor);
    }

    public static Vec3f divide(Vec3f v, float divisor) {
        if (DataHelper.isFloatEqual(divisor, 1.0f)) {
            return copy(v);
        }
        return new Vec3f(v.getX() / divisor, v.getY() / divisor, v.getZ() / divisor);
    }

    public static Vec3i getOffset(BlockPos begin, BlockPos end) {
        return new Vec3i(end.getX() - begin.getX(), end.getY() - begin.getY(), end.getZ() - begin.getZ());
    }

    public static BlockPos applyOffset(BlockPos v, Vec3i offset) {
        return new BlockPos(v.getX() + offset.getX(), v.getY() + offset.getY(), v.getZ() + offset.getZ());
    }

    public static TargetBlockPos applyOffset(TargetBlockPos v, Vec3i offset) {
        return new TargetBlockPos(v.world(), v.getX() + offset.getX(), v.getY() + offset.getY(),
                v.getZ() + offset.getZ());
    }

    public static Vec3f applyOffset(Vec3f v, Vec3i offset) {
        return new Vec3f(v.getX() + offset.getX(), v.getY() + offset.getY(), v.getZ() + offset.getZ());
    }

    public static Vec3f applyOffset(Vec3f v, Vec3f offset) {
        return new Vec3f(v.getX() + offset.getX(), v.getY() + offset.getY(), v.getZ() + offset.getZ());
    }

    public static Vec3f getPerpendicularVector(Direction v) {
        return new Vec3f(1 - v.getOffsetX(), 1 - v.getOffsetY(), 1 - v.getOffsetZ());
    }

    public static String toString(BlockPos v) {
        return "%s, %s, %s".formatted(v.getX(), v.getY(), v.getZ());
    }

    public static String toString(TargetBlockPos v) {
        return v.toString();
    }
}

File: src/main/java/me/nk0/rcu/helper/PosHelper.kt
package me.nk0.rcu.helper

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import me.nk0.rcu.util.Blocks
import me.nk0.rcu.util.TargetBlockPos
import me.nk0.rcu.util.Vec3f
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.math.Vec3i
import org.joml.Vector3f

object PosHelper {
    @JvmStatic
    fun toVec3i(v: BlockPos?): Vec3i? {
        if (v == null) {
            return null
        }
        return Vec3i(v.x, v.y, v.z)
    }

    @JvmStatic
    fun toBlockPos(v: Vec3i?): BlockPos? {
        if (v == null) {
            return null
        }
        return BlockPos(v)
    }

    fun toVec3i(v: Vec3f?): Vec3i? {
        if (v == null) {
            return null
        }
        return Vec3i(v.getX().toInt(), v.getY().toInt(), v.getZ().toInt())
    }

    fun toVec3f(v: Vec3i?): Vec3f? {
        if (v == null) {
            return null
        }
        return Vec3f(v.x.toFloat(), v.y.toFloat(), v.z.toFloat())
    }

    fun toVec3f(v: Vec3f): Vec3f {
        return v
    }

    fun toVec3f(v: Vector3f?): Vec3f {
        return Vec3f(v)
    }

    fun fromJson(v: JsonElement): Vec3i {
        return fromJson(v.asJsonArray)
    }

    @JvmStatic
    fun fromJson(v: JsonArray): Vec3i {
        return Vec3i(v[0].asInt, v[1].asInt, v[2].asInt)
    }

    fun center(v: BlockPos): Vec3f {
        return Vec3f(v.x + 0.5f, v.y + 0.5f, v.z + 0.5f)
    }

    @JvmStatic
    fun copy(v: BlockPos?): BlockPos? {
        if (v == null) {
            return null
        }
        return v.mutableCopy()
    }

    fun copy(v: TargetBlockPos?): TargetBlockPos? {
        if (v == null) {
            return null
        }
        return v.copy()
    }

    @JvmStatic
    fun copy(v: Vec3i?): Vec3i? {
        if (v == null) {
            return null
        }
        return Vec3i(v.x, v.y, v.z)
    }

    fun copy(v: Vec3f?): Vec3f? {
        if (v == null) {
            return null
        }
        return Vec3f(v.getX(), v.getY(), v.getZ())
    }

    @JvmStatic
    fun copy(v: Blocks?): Blocks? {
        if (v == null) {
            return null
        }
        return Blocks(v.first(), v.increment(), v.size())
    }

    @JvmStatic
    fun scale(v: Vec3i, factor: Int): Vec3i? {
        if (factor == 1) {
            return copy(v)
        }
        return v.multiply(factor)
    }

    @JvmStatic
    fun divide(v: Vec3i, divisor: Int): Vec3i? {
        if (divisor == 1) {
            return copy(v)
        }
        return Vec3i(v.x / divisor, v.y / divisor, v.z / divisor)
    }

    fun scale(v: Vec3f?, factor: Float): Vec3f? {
        if (DataHelper.isFloatEqual(factor, 1.0f)) {
            return copy(v)
        }
        return Vec3f(v!!.getX() * factor, v.getY() * factor, v.getZ() * factor)
    }

    fun divide(v: Vec3f, divisor: Float): Vec3f? {
        if (DataHelper.isFloatEqual(divisor, 1.0f)) {
            return copy(v)
        }
        return Vec3f(v.getX() / divisor, v.getY() / divisor, v.getZ() / divisor)
    }

    @JvmStatic
    fun getOffset(begin: BlockPos, end: BlockPos): Vec3i {
        return Vec3i(end.x - begin.x, end.y - begin.y, end.z - begin.z)
    }

    @JvmStatic
    fun applyOffset(v: BlockPos, offset: Vec3i): BlockPos {
        return BlockPos(v.x + offset.x, v.y + offset.y, v.z + offset.z)
    }

    fun applyOffset(v: TargetBlockPos, offset: Vec3i): TargetBlockPos {
        return TargetBlockPos(v.world(), v.x + offset.x, v.y + offset.y,
                v.z + offset.z)
    }

    fun applyOffset(v: Vec3f, offset: Vec3i): Vec3f {
        return Vec3f(v.getX() + offset.x, v.getY() + offset.y, v.getZ() + offset.z)
    }

    fun applyOffset(v: Vec3f?, offset: Vec3f?): Vec3f {
        return Vec3f(v!!.getX() + offset!!.getX(), v.getY() + offset.getY(), v.getZ() + offset.getZ())
    }

    fun getPerpendicularVector(v: Direction): Vec3f {
        return Vec3f((1 - v.offsetX).toFloat(), (1 - v.offsetY).toFloat(), (1 - v.offsetZ).toFloat())
    }

    fun toString(v: BlockPos): String {
        return "%s, %s, %s".format(v.x, v.y, v.z)
    }

    @JvmStatic
    fun toString(v: TargetBlockPos): String {
        return v.toString()
    }
}
